text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,526 @@
+
+Instruments that are probably not seen as widely useful:
+
+- _UpDownCumulativeCounter_: We believe this is better handled asynchronously.
+
+#### Observer instruments
+
+The foundational `Observer` instrument without refinements or
+restrictions shall be called a `LastValueObserver` instrument.
+
+We have identified important cases that should be standardized:
+
+- _CumulativeObserver_: Support a cumulative monotone counter
+- _DeltaObserver_: Support an asynchronous delta counter.
+
+Observer refinements that could be standardized in the future:
+
+- _UpDownCumulativeObserver_: Observe a non-monotonic cumluative counter
+- _UpDownDeltaObserver_: Observe positive and negative deltas
+- _AbsoluteLastValueObserver_: Observe non-negative current values.
+
+## Example: Observer aggregation
+
+Suppose you wish to capture the CPU usage of a process broken down by
+the CPU core ID.  The operating system provides a mechanism to read
+the current usage from the `/proc` file system, which will be reported
+once per collection interval using an Observer instrument.  Because
+this is a precomputed sum with a non-negative rate, use a
+_CumulativeObserver_ to report this quantity with a metric label
+indicating the CPU core ID.
+
+It will be common to compute a rate of CPU usage over this data.  The
+rate can be calculated for an individual CPU core by computing a
+difference between the value of two metric events.  To compute the
+aggregate rate across all cores–a spatial aggregation–these
+differences are added together.
+
+## Open Questions
+
+Are there still questions surrounding the former Monotonic refinement?
+
+Should the _CumulativeObserver_ instrument be named
+_MonotonicObserver_?  In this proposal, we prefer _Cumulative_ and
+_UpDownCumulative_.  _Cumulative_ is a good descriptive term in this
+setting (i.e., some additive values are _cumulative_, some are
+_delta_).  Being _Cumulative_ and not _UpDownCumulative_ implies
+monotonicity in this proposal.
+
+For synchronous instruments, this proposals does not standardize
+_CumulativeCounter_. Such an instrument might be named
+_MonotonicCounter_.
+
+## Trade-offs and mitigations
+
+The trade-off explicitly introduced here is that we should prefer to
+create new instrument refinements, each for a dedicated purpose,
+rather than create generic instruments with support for multiple
+semantic options.
+
+## Prior art and alternatives
+
+The optional behaviors `Monotonic` and `Absolute` were first discussed
+in the August 2019 Metrics working group meeting.
+
+## Future possibilities
+
+A future OTEP will request the introduction of two standard
+refinements for the 0.4 API specification.  This will be the
+`CumulativeObserver` instrument described above plus a synchronous
+timing instrument named `TimingMeasure` that is equivalent to
+_AbsoluteMeasure_ with the correct unit and a language-specific
+duration type for measuring time.
+
+If the above open question is decided in favor of treating the
```

Comment:
I think the open question referenced disappeared in some edit. It's not there now.

Reply:
The question was whether names like "Measure" and "Observer" should be reserved as explanatory, descriptive names, and whether we should find new, different names for the actual, concrete unrestricted/unrefined form of these instruments handled by users. This was discussed in the SIG call but no issue specifically was filed. 

This has been referred to as "Abstract foundation instruments" vs "Non-abstract foundation instruments". I originally proposed Non-abstract, i.e., that the unrestricted Measure instrument should be called Measure. @bogdandrutu convinced me to go the other way, which I have done in this document's sample proposal and in OTEP 93. 

However, I have changed that stance again in OTEP 96, which states that we should use "Recorder" instead of "Measure", and that the default kind of Recorder instance should be named "Recorder--this is the Non-abstract proposal.

Reply:
oh, I know the open question...it just isn't in this doc any more that I could see.  :)